feat(Kahoot): Add new paths and fix bugs

theusaf opened the pull request:
**Describe the changes in this pull request:**
* Fixes issues with the Kahoot! presence not working for many areas. 
* Added new paths and updated old ones to reflect the current Kahoot! application
* Added details for the create.kahoot.it page to the presence

<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<p><em>Screenshots are shown without the string translations implemented yet.</em></p>
<br>
<!-- Required when a presence has been added or modified --->
<!-- Attach screenshot(s) here --->
<img width="150" alt="2022-03-25 16_03_17-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219825-fbc7a0ca-3e63-4db4-8b5d-4bba557042c9.png">
<img width="150" alt="2022-03-25 16_03_32-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219826-6394593c-8d2f-49fc-83a3-6b819cfddff2.png">
<img width="152" alt="2022-03-25 16_03_44-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219827-1fd48fdd-1ba4-4470-ac23-8d427248e431.png">
<img width="148" alt="2022-03-25 16_03_52-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219828-9bdb1c4a-c64a-4885-93cd-5e837e253ab2.png">
<img width="149" alt="2022-03-25 16_03_57-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219829-715c4b21-04e7-48df-b0ee-51a35141f081.png">
<img width="150" alt="2022-03-25 16_04_03-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219831-08b5abbb-48b1-48f7-a22a-6aa8e031ce5c.png">
<img width="153" alt="2022-03-25 16_04_11-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219834-a0442e71-b6e2-49a4-8641-199210d7a15c.png">
<img width="148" alt="2022-03-25 16_04_46-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219835-5fa0aee1-dd64-4e13-9e9c-040aba0a7135.png">
<img width="147" alt="2022-03-25 16_27_08-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219837-83726006-8585-4fff-8ba2-be443b5e68a3.png">
<img width="147" alt="2022-03-25 16_31_31-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219838-0e54cd61-b2b0-4e84-800e-d8a0f0bc6f9c.png">
<img width="146" alt="2022-03-25 16_33_29-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219839-020af56e-47a2-4e6c-9b01-107725965a44.png">
<img width="150" alt="2022-03-25 16_33_42-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219840-99a0f5a5-98f6-4ef8-84dd-5e085c6754ea.png">
<img width="148" alt="2022-03-25 16_54_07-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219841-b096c8ca-9661-41ca-a779-0be0f5ae87a1.png">
<img width="151" alt="2022-03-25 16_54_30-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219842-df275fc1-7ef7-4429-95cf-ea1d8c5570f8.png">
<img width="141" alt="2022-03-25 16_57_08-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219843-39d919c2-1065-4123-8e58-968589a1f5de.png">
<img width="147" alt="2022-03-25 16_57_19-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219844-051f668d-6885-4482-9ba9-c0b61bbbfefd.png">
<img width="145" alt="2022-03-25 17_11_05-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219845-72da1f99-1a93-41f7-92e3-fb25bb72bf6f.png">
<img width="148" alt="2022-03-25 17_11_12-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219846-d86dac26-1721-45d1-a40a-4a750ef46d5a.png">
<img width="148" alt="2022-03-25 17_14_28-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219847-ae500c01-0c74-4e50-86d0-00d7b5b4f3bf.png">
<img width="147" alt="2022-03-25 17_16_21-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219850-675af84f-a83e-47ed-a024-cbb2ba997fec.png">
<img width="146" alt="2022-03-25 17_16_59-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219852-4feb4cb6-5aca-48bd-8dcb-5eaac6727f90.png">
<img width="144" alt="2022-03-25 17_19_52-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219853-7a0d93a4-0993-49fc-96d0-843b475fd926.png">
<img width="143" alt="2022-03-25 17_20_21-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219854-b6303188-c33e-4773-ac54-30d6aeb648bd.png">
<img width="146" alt="2022-03-25 18_15_30-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219855-6686857d-ddc5-4c12-971a-2d0fe586d2a5.png">
<img width="146" alt="2022-03-25 18_15_38-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219856-e52b361a-1ef5-4a3f-9529-ad0cead9a374.png">
<img width="147" alt="2022-03-25 18_15_46-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219857-e2496821-746d-4ea7-90ec-23d8c2c72688.png">
<img width="147" alt="2022-03-25 18_16_49-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219858-7d7f5cd1-e10d-4e42-a7ef-57ee31d69422.png">
<img width="146" alt="2022-03-25 18_16_59-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219859-791a569e-9350-40e7-b0fb-222cd7ef06dc.png">
<img width="149" alt="2022-03-25 18_17_10-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219860-3395ac2d-09f4-4827-b386-856beb0cebea.png">
<img width="148" alt="2022-03-25 18_17_30-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219861-1251366e-3af1-49c8-9bc3-682a27051f4d.png">
<img width="146" alt="2022-03-25 18_17_48-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219863-46c26e90-dfd7-45a3-894a-e07ff5370b6e.png">
<img width="144" alt="2022-03-25 18_18_27-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219864-a898810e-36dc-4214-9558-a1d2476a8496.png">
<img width="149" alt="2022-03-25 18_18_49-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219865-0d928bcf-0211-45d6-bdbe-df2ab978f698.png">
<img width="148" alt="2022-03-25 18_19_17-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219866-96668f7e-bf0f-4e7a-a024-32fb0570fea1.png">
<img width="146" alt="2022-03-25 18_21_07-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219867-d1ccd412-c921-46e2-9121-04daba1efd49.png">
<img width="148" alt="2022-03-25 18_30_53-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219868-7d1f74af-3ae1-4dc2-865b-0957955c1d67.png">
<img width="149" alt="2022-03-25 18_31_00-presence-dev - Discord" src="https://user-images.githubusercontent.com/32113157/160219869-8c50a131-7c77-4e9f-b5e5-8665b9d3e8a6.png">

</details>

theusaf committed:
fix(kahoot.it): Fix scraping methods

- Uses [data-functional-selector], which changes less often than the class names
- Added some strings to cover:
  - feedback
  - waiting on answer
  - poll questions

theusaf committed:
feat(kahoot.it): Add new paths

- Also fixes some issues with certain messages

theusaf committed:
fix(play.kahoot.it): Fix scraping method

- Combines /contentblock and /gameblock
  - They are both under /gameblock, so a different method is used to determine if it is a slide or not.

theusaf committed:
feat(play.kahoot.it): Add game-summary path

theusaf committed:
refactor(play.kahoot.it): Use switch case for play.kahoot.it paths

theusaf committed:
refactor: Remove unessessary global variables

theusaf committed:
feat(create.kahoot.it): Add presence for create.kahoot.it

theusaf committed:
style: Add and remove whitespace where needed

theusaf committed:
chore: bump version and add contributor

2.0.10 --> 2.1.0
Added myself as a contributor as I believe this should constitute >25% changes

theusaf committed:
fix(findRanking): Should not be an async function

theusaf committed:
refactor: Remove unnessessary if statement

theusaf committed:
chore: Update comments

theusaf commented on websites/K/Kahoot/dist/metadata.json:
```diff
@@ -22,9 +26,10 @@
+		"create.kahoot.it"
 	],
-	"version": "2.0.10",
+	"version": "2.1.0",
```
Would this PR be considered a minor or major update?

EncryptedDev replied:
Breaking

theusaf committed:
feat: Add elapsed timestamps

theusaf committed:
perf: Update timestamp immediately

theusaf committed:
perf: Remove button when game is locked

theusaf committed:
perf: lint

EncryptedDev commented on websites/K/Kahoot/presence.ts:
```diff
@@ -1,61 +1,84 @@
+			correctAnswer: "kahoot.correctAnswer", // Correct Answer
+			pollAnswer: "kahoot.pollAnswer", // + Answered Poll
+			resultsQuestion: "kahoot.resultsQuestion", // Looking at Results:
+			slideViewing: "kahoot.slideViewing", // Viewing Slide with Content...
+			gameOver: "kahoot.gameOver", // Game Over
+			gameCreate: "kahoot.gameCreate", // Creating a Game...
+			loadingPage: "kahoot.loadingPage", // Loading Page...
+			firstPlace: "kahoot.firstPlace", // First Place: {0}
+			points: "kahoot.points", // Points: {0}
+			questionsCorrect: "kahoot.questionsCorrect", // Questions Correct: {0}
+			slideShowing: "kahoot.slideShowing", // Showing Slide with Content
+			questionShowing: "kahoot.questionShowing", // Showing Question:
+			stString: "kahoot.stString", // 1st place
+			ndString: "kahoot.ndString", // 2nd place
+			rdString: "kahoot.rdString", // 3rd place
+			topX: "kahoot.topX", // Top {0}
+			onPodium: "kahoot.onPodium", // On the Podium
+			of: "kahoot.of", // {0} of {1}
+			questionNumber: "kahoot.questionNumber", // Question: {0}
+			feedback: "kahoot.feedback", // Providing feedback
+			waitingAnswer: "kahoot.waitingAnswer", // Waiting for results...
+			drumRoll: "kahoot.drumRoll", // Waiting for final ranking...
+			position: "kahoot.position", // Position: {0}
+			teamTalk: "kahoot.teamTalk", // Discussing with team
+			gameSummary: "kahoot.gameSummary", // Looking at game summary
+			login: "kahoot.login", // Logging in
+			createHome: "kahoot.createHome", // Viewing Create home Page
+			discover: "kahoot.discover", // Viewing Kahoot! discover Page
+			searchKahoots: "kahoot.searchKahoots", // Searching for Kahoots:
+			kahootDetails: "kahoot.kahootDetails", // Viewing Kahoot details:
+			kahootProfile: "kahoot.kahootProfile", // Viewing Kahoot profile:
+			myKahoots: "kahoot.myKahoots", // Viewing their Kahoots
+			userReports: "kahoot.userReports", // Viewing game reports
+			myCourses: "kahoot.myCourses", // Viewing their courses
+			editingCourse: "kahoot.editingCourse", // Editing a course:
+			viewingCourse: "kahoot.viewingCourse", // Viewing a course:
+			editingKahoot: "kahoot.editingKahoot", // Editing a Kahoot:
+			previewingKahoot: "kahoot.previewingKahoot", // Previewing a Kahoot
+			liveCourse: "kahoot.liveCourse", // Conducting a live course
+			liveCourseActivity: "kahoot.liveCourseActivity" // Activity #{0}: {1}
```
Comments like such are not used anywhere else in the repository so should not be used here either. A future maintainer can reference the strings from PreMiD/Localization without much inconvenience.

EncryptedDev commented on websites/K/Kahoot/presence.ts:
```diff
@@ -1,61 +1,84 @@
+			liveCourseActivity: "kahoot.liveCourseActivity" // Activity #{0}: {1}
 		},
-		await presence.getSetting<string>("lang")
+		await presence.getSetting("lang")
```
Why was the type removed?

EncryptedDev commented on websites/K/Kahoot/presence.ts:
```diff
@@ -1,61 +1,84 @@
 	oldLang: string = null,
-	rankingSelector: HTMLElement;
+	browsingTimestamp = Math.floor(Date.now() / 1000),
+	/**
+	 * 0 - ready to be updated if needed
+	 * 1 - updated, ready to be reset to 0
+	 */
```
Use an inline comment instead, these comments are not used for this purpose

EncryptedDev commented on websites/K/Kahoot/presence.ts:
```diff
@@ -1,61 +1,84 @@
+			largeImageKey: "kahoot",
+			startTimestamp: browsingTimestamp
 		},
 		buttons = await presence.getSetting<boolean>("buttons"),
 		newLang = await presence.getSetting<string>("lang");
```
Use Promise.all

EncryptedDev commented on websites/K/Kahoot/presence.ts:
```diff
@@ -64,127 +87,323 @@ presence.on("UpdateData", async () => {
 	if (oldLang !== newLang) oldLang = newLang;
 
-	if (document.location.href.match("https://kahoot.it")) {
-		const path = document.location.pathname;
```
You should keep this and destructure `document.location.pathname` and `document.location.host` for simplicity

theusaf committed:
refactor: Fix comments and other issues

Removed comments on strings as requested
Used destructuring on the document.location to simplify code
Used promise.all on getting settings to make code cleaner and faster

theusaf committed:
chore: bump version

Breaking change
--> 3.0.0

ImRodry commented on websites/K/Kahoot/dist/metadata.json:
```diff
@@ -22,9 +26,10 @@
+		"create.kahoot.it"
 	],
-	"version": "2.0.10",
+	"version": "3.0.0",
```
please explain why this is breaking

theusaf replied:
@EncryptedDev told me that this was breaking.
https://github.com/PreMiD/Presences/pull/5975#discussion_r835824127

Some reasons this could be breaking is that it adds a new domain, and a lot of the original code has been rewritten.

ImRodry replied:
Adding a new domain is minor not breaking. Recoding is also not breaking unless you remove support for something. @EncryptedDev you should also explain comments like that a bit better

EncryptedDev replied:
I assumed a bunch of routes were removed, didn’t see they were replaced by the switch-case

ImRodry commented on websites/K/Kahoot/presence.ts:
```diff
@@ -38,153 +37,378 @@ async function getStrings() {
-	ranking: string,
-	top5: Promise<boolean>,
-	strings: { [key: string]: string },
+let strings: { [key: string]: string },
```
this type is wrong, it should be more specific

theusaf replied:
This was from the original codebase. What do you suggest I change it to, or how can I make it more specific?

ImRodry replied:
Follow the standard in other popular presences by using ReturnType

ImRodry commented on websites/K/Kahoot/presence.ts:
```diff
@@ -38,153 +37,378 @@ async function getStrings() {
 
 	oldLang ??= newLang;
 	strings ??= await getStrings();
 	if (oldLang !== newLang) oldLang = newLang;
```
where do the strings get updated?

theusaf replied:
This was also from the original codebase. I'll look into fixing this problem.

theusaf committed:
revert: chore: bump version

theusaf committed:
perf(strings): make typing more specific

theusaf committed:
fix: update strings upon language option change

Slowlife01 committed:
Merge branch 'main' into kahoot-fix

Slowlife01 committed:
Merge branch 'main' into kahoot-fix

Slowlife01 commented:
Please resolve eslint issues

theusaf committed:
refactor: run linter

theusaf committed:
feat: add kahoot hosting details to preview

The reason the iframe.ts file is mostly in a block is to allow it to actually
compile. tsc thinks that it can access things in presence.ts, which it cannot.

theusaf committed:
fix: update logo and colors

theusaf committed:
chore: add hostname to create home page state

theusaf committed:
revert: feat: add kahoot hosting details to preview

Instead adds the name of the kahoot being previewed

theusaf committed:
style: combine let statements

Bas950 commented on websites/K/Kahoot/presence.ts:
```diff
@@ -1,18 +1,16 @@
-async function findRanking(rankingSelector: HTMLElement) {
-	if (
+const presence = new Presence({
+	clientId: "958898877598146571"
```
please explain why a new clientid was needed.

theusaf replied:
To update the large image

ImRodry replied:
You can contact @StrikerFRFX about that

theusaf replied:
Got it.

theusaf replied:
What is the best way to contact them?

Bas950 replied:
Idk probably discord

EncryptedDev replied:
Has the assets been added now?

theusaf replied:
I have tried contacting them, but they haven't responded yet.

theusaf replied:
The assets have been added.
<img width="220" alt="2022-04-01 07_24_45" src="https://user-images.githubusercontent.com/32113157/161283147-b42b8073-de05-4017-b963-5ca06b243929.png">

theusaf committed:
revert: fix: update logo and colors

Reverts the clientId, will contact original owner for changes
